// pikola/lib/functions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EveryMonth = exports.EveryDay = exports.EveryHour = exports.EveryMinute = exports.EverySecond = exports.EveryMillisecond = exports.OnWeek = exports.OnDayOfMonth = exports.EveryWeek = exports.OnDayOfWeek = exports.OnHour = exports.OnMinute = exports.OnSecond = exports.OnMillisecond = undefined;

var _scheduler = require('./scheduler');

var _resolution = require('./resolution');

var _resolution2 = _interopRequireDefault(_resolution);

var _recurrence = require('./recurrence');

var _recurrence2 = _interopRequireDefault(_recurrence);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function func(resolution, type, validate, beforeRun) {
  return function (interval) {
    if (validate) validate(interval);
    return {
      resolution: resolution,
      execute: function execute(opt) {
        if (beforeRun) opt = beforeRun.execute(opt);
        return (0, _scheduler.addLayer)({
          resolution: resolution,
          type: type,
          interval: interval
        }, opt);
      }
    };
  };
}

var OnMillisecond = exports.OnMillisecond = func(_resolution2.default.Millisecond, _recurrence2.default.On, function (interval) {
  if (interval < 0) throw Error('Interval in OnMillisecond can\'t be lower than 0');
  if (interval > 999) throw Error('Interval in OnMillisecond can\'t be higher than 999');
});

var OnSecond = exports.OnSecond = func(_resolution2.default.Second, _recurrence2.default.On, function (interval) {
  if (interval < 0) throw Error('Interval in OnSecond can\'t be lower than 0');
  if (interval > 59) throw Error('Interval in OnSecond can\'t be higher than 59');
}, OnMillisecond(0));

var OnMinute = exports.OnMinute = func(_resolution2.default.Minute, _recurrence2.default.On, function (interval) {
  if (interval < 0) throw Error('Interval in OnMinute can\'t be lower than 0');
  if (interval > 59) throw Error('Interval in OnMinute can\'t be higher than 59');
}, OnSecond(0));

var OnHour = exports.OnHour = func(_resolution2.default.Hour, _recurrence2.default.On, function (interval) {
  if (interval < 0) throw Error('Interval in OnHour can\'t be lower than 0');
  if (interval > 23) throw Error('Interval in OnHour can\'t be higher than 23');
}, OnMinute(0));

var OnDayOfWeek = exports.OnDayOfWeek = func(_resolution2.default.Day, _recurrence2.default.On, function (interval) {
  if (interval < 0) throw Error('Interval in OnDayOfWeek can\'t be lower than 0');
  if (interval > 6) throw Error('Interval in OnDayOfWeek can\'t be higher than 6');
}, OnHour(0));

var EveryWeek = exports.EveryWeek = func(_resolution2.default.Week, _recurrence2.default.Every, function (interval) {
  if (interval < 1) throw Error('Interval in EveryWeek can\'t be lower than 1');
}, OnDayOfWeek(0));

var OnDayOfMonth = exports.OnDayOfMonth = func(_resolution2.default.Day, _recurrence2.default.On, function (interval) {
  if (interval < 0) throw Error('Interval in OnDayOfMonth can\'t be lower than 0');
  if (interval > 30) throw Error('Interval in OnDayOfMonth can\'t be higher than 30');
}, func(_resolution2.default.Week, _recurrence2.default.Every, null, OnDayOfWeek(0))(null));

var OnWeek = exports.OnWeek = func(_resolution2.default.Week, _recurrence2.default.On, function (interval) {
  if (interval < 0) throw Error('Interval in OnWeek can\'t be lower than 0');
  if (interval > 4) throw Error('Interval in OnWeek can\'t be higher than 4');
}, OnDayOfWeek(0));

var EveryMillisecond = exports.EveryMillisecond = func(_resolution2.default.Millisecond, _recurrence2.default.Every, function (interval) {
  if (interval < 1) throw Error('Interval in EveryMillisecond can\'t be lower than 1');
});

var EverySecond = exports.EverySecond = func(_resolution2.default.Second, _recurrence2.default.Every, function (interval) {
  if (interval < 1) throw Error('Interval in EverySecond can\'t be lower than 1');
}, OnMillisecond(0));

var EveryMinute = exports.EveryMinute = func(_resolution2.default.Minute, _recurrence2.default.Every, function (interval) {
  if (interval < 1) throw Error('Interval in EveryMinute can\'t be lower than 1');
}, OnSecond(0));

var EveryHour = exports.EveryHour = func(_resolution2.default.Hour, _recurrence2.default.Every, function (interval) {
  if (interval < 1) throw Error('Interval in EveryHour can\'t be lower than 1');
}, OnMinute(0));

var EveryDay = exports.EveryDay = func(_resolution2.default.Day, _recurrence2.default.Every, function (interval) {
  if (interval < 1) throw Error('Interval in EveryDay can\'t be lower than 1');
}, OnHour(0));

var EveryMonth = exports.EveryMonth = func(_resolution2.default.Month, _recurrence2.default.Every, function (interval) {
  if (interval < 1) throw Error('Interval in EveryMonth can\'t be lower than 1');
}, OnDayOfMonth(0));
